File: android/core/src/test/kotlin/io/github/kegustafsson/driveremote/core/SourcesTest.kt
package io.github.kegustafsson.driveremote.core

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource

/** Ported one-for-one from `sk-plugin/src/pure/sources.test.ts`. */
class SourcesTest {

  @ParameterizedTest
  @MethodSource("validIds")
  fun `accepts the exact id`(raw: String, expected: CommandSource) {
    assertEquals(expected, parseSource(raw))
  }

  @ParameterizedTest
  @MethodSource("garbage")
  fun `maps unrecognised readings to unknown`(raw: Any?) {
    assertEquals(CommandSource.UNKNOWN, parseSource(raw))
  }

  @Test
  fun `gives human labels for each id`() {
    assertEquals("local switch", sourceLabel("local"))
    assertEquals("TX remote", sourceLabel("tx"))
    assertEquals("this app", sourceLabel("plugin"))
    assertEquals("nobody", sourceLabel("none"))
    assertEquals("unknown", sourceLabel("garbage"))
  }

  @Test
  fun `local and tx outrank this app`() {
    // Fixed precedence, never recency-based (AGENTS.md non-negotiable 3).
    assertTrue(overridesThisApp("local"))
    assertTrue(overridesThisApp("tx"))
  }

  @Test
  fun `plugin, none and unknown do not outrank this app`() {
    assertFalse(overridesThisApp("plugin")) // that's us
    assertFalse(overridesThisApp("none"))
    assertFalse(overridesThisApp("whatever"))
  }

  @Test
  fun `wire strings match the telemetry contract exactly`() {
    // Pinned as literals: these come off the wire from RX and HH, so a rename
    // would silently stop the "controlled by ..." note from ever appearing.
    assertEquals("local", CommandSource.LOCAL.wire)
    assertEquals("tx", CommandSource.TX.wire)
    assertEquals("plugin", CommandSource.PLUGIN.wire)
    assertEquals("none", CommandSource.NONE.wire)
  }

  companion object {
    @JvmStatic
    fun validIds() =
      listOf(
        org.junit.jupiter.params.provider.Arguments.of("local", CommandSource.LOCAL),
        org.junit.jupiter.params.provider.Arguments.of("tx", CommandSource.TX),
        org.junit.jupiter.params.provider.Arguments.of("plugin", CommandSource.PLUGIN),
        org.junit.jupiter.params.provider.Arguments.of("none", CommandSource.NONE),
      )

    /** Mirrors the vitest list `[undefined, null, '', 'TX', 'server', 42, {}]`. */
    @JvmStatic
    fun garbage(): List<Any?> = listOf(null, "", "TX", "server", 42, emptyMap<String, Any>())
  }
}

File: sk-plugin/src/pure/sources.test.ts
import { describe, expect, it } from 'vitest';
import { overridesThisApp, parseSource, sourceLabel } from './sources';

describe('parseSource', () => {
  it.each(['local', 'tx', 'plugin', 'none'] as const)(
    'accepts the exact id %s',
    (value) => {
      expect(parseSource(value)).toBe(value);
    },
  );

  it.each([undefined, null, '', 'TX', 'server', 42, {}])(
    'maps unrecognised %p to unknown',
    (value) => {
      expect(parseSource(value)).toBe('unknown');
    },
  );
});

describe('sourceLabel', () => {
  it('gives human labels for each id', () => {
    expect(sourceLabel('local')).toBe('local switch');
    expect(sourceLabel('tx')).toBe('TX remote');
    expect(sourceLabel('plugin')).toBe('this app');
    expect(sourceLabel('none')).toBe('nobody');
    expect(sourceLabel('garbage')).toBe('unknown');
  });
});

describe('overridesThisApp', () => {
  it('local and tx outrank this app', () => {
    expect(overridesThisApp('local')).toBe(true);
    expect(overridesThisApp('tx')).toBe(true);
  });

  it('plugin (us), none, and unknown do not', () => {
    expect(overridesThisApp('plugin')).toBe(false);
    expect(overridesThisApp('none')).toBe(false);
    expect(overridesThisApp('whatever')).toBe(false);
  });
});
